### ISSUE TITLE
accelerate does not load from the checkpoint properly

### ISSUE THREAD

**Issue by wilnn** [huggingface/accelerate#3747]
I am new to accelerate, and I am testing the checkpointing stuff. I have this test script which basically testing if dataloader wraped by acclerator will produce the same shuffled dataset order everytime run it and from checkpoint:
```python
from torch.utils.data import DataLoader, DistributedSampler
import torch.nn as nn
from accelerate import Accelerator
import torch
from torch.utils.data import Dataset, DataLoader
import torch.nn as nn, torch.optim as optim
from accelerate.utils import DistributedDataParallelKwargs, DistributedType, ProjectConfiguration, set_seed
from torch.optim import lr_scheduler

set_seed(42)

class MyDataset(Dataset):
    def __init__(self, n):
        self.data = list(range(n))
    def __len__(self):
        return len(self.data)
    def __getitem__(self, idx):
        return torch.tensor(self.data[idx])

dataset = MyDataset(10)

dataloader = DataLoader(dataset, batch_size=2, shuffle=True)

class PseudoModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(1, 1)
    def forward(self, x):
        return self.fc(x)

model = PseudoModel()
optimizer = optim.SGD(model.parameters(), lr=0.01)
scheduler = lr_scheduler.StepLR(optimizer, step_size=30, gamma=0.1)

accelerator_project_config = ProjectConfiguration(project_dir="test")
kwargs = DistributedDataParallelKwargs(find_unused_parameters=True)
accelerator = Accelerator(project_dir="my/", kwargs_handlers=[kwargs])

mse_loss_fn = nn.MSELoss()
model, optimizer, dataloader, scheduler = accelerator.prepare(model, optimizer, dataloader, scheduler)

for epoch in range(2):
    for step, batch in enumerate(dataloader):
        print(f"Epoch {epoch}, Step {step}, Batch {batch}")
        output = model(torch.tensor([1.0], device='cuda:0'))
        loss = mse_loss_fn(output, torch.tensor([1.0], device='cuda:0'))
        accelerator.backward(loss)
        optimizer.step()
        scheduler.step()
        optimizer.zero_grad()
        if epoch == 1 and step == 2:
            accelerator.save_state("checkpoint-2")
            exit()
```

everytime I run it, The dataloader in the script above can shuffle the dataset to the same order  (the dataset is shuffle, but it is the same shuffled order everytime). This script will use accelerate to save the checkpoint.
Then, I use this script which is the same as the one above except that it will load from the checkpoint that is saved above:
```python
from torch.utils.data import DataLoader, DistributedSampler
import torch.nn as nn
from accelerate import Accelerator
import torch
from torch.utils.data import Dataset, DataLoader
import torch.nn as nn, torch.optim as optim
from accelerate.utils import DistributedDataParallelKwargs, DistributedType, ProjectConfiguration, set_seed
from torch.optim import lr_scheduler

set_seed(42)

class MyDataset(Dataset):
    def __init__(self, n):
        self.data = list(range(n))
    def __len__(self):
        return len(self.data)
    def __getitem__(self, idx):
        return torch.tensor(self.data[idx])

dataset = MyDataset(10)

dataloader = DataLoader(dataset, batch_size=2, shuffle=True)

class PseudoModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(1, 1)
    def forward(self, x):
        return self.fc(x)

model = PseudoModel()
optimizer = optim.SGD(model.parameters(), lr=0.01)
scheduler = lr_scheduler.StepLR(optimizer, step_size=30, gamma=0.1)

accelerator_project_config = ProjectConfiguration(project_dir="test")
kwargs = DistributedDataParallelKwargs(find_unused_parameters=True)
accelerator = Accelerator(project_dir="my/", kwargs_handlers=[kwargs])

mse_loss_fn = nn.MSELoss()
model, optimizer, dataloader, scheduler = accelerator.prepare(model, optimizer, dataloader, scheduler)

accelerator.load_state("checkpoint-2")

for epoch in range(2):
    for step, batch in enumerate(dataloader):
        print(f"Epoch {epoch}, Step {step}, Batch {batch}")
        output = model(torch.tensor([1.0], device='cuda:0'))
        loss = mse_loss_fn(output, torch.tensor([1.0], device='cuda:0'))
        accelerator.backward(loss)
        optimizer.step()
        scheduler.step()
        optimizer.zero_grad()
```
Somehow, after loading from the above checkpoint, the dataloader produce a different shuffled order of the dataset as the first script . Running the second script multiple times produce the same shuffled order of the dataset but that shuffle ordered is different from the one in the second script while they should be the same. I think the `accelerator.load_state()` mess things up. I don't know if I have don't anything wrong here. Please help. Thank you

**Comment by MengAiDev** [comment]
Thanks for sharing this detailed question! This is actually a common challenge when working with distributed training and checkpointing, so you're definitely not alone in running into this.

From what you've described, it seems like the issue comes down to how the distributed sampler manages its internal state. When you use `accelerator.prepare()` on your DataLoader, Accelerate replaces it with a version that uses a `DistributedSampler` behind the scenes. This sampler has its own random state that gets saved and loaded with checkpoints.

The problem likely occurs because:
1. Your initial `set_seed(42)` gets called before the checkpoint load
2. The loaded sampler state might not be perfectly synchronized with Python's random state
3. The sampler needs to know which epoch you're on to maintain consistent shuffling

Here's what usually works for me in these situations:

```python
# After loading your checkpoint
accelerator.load_state("checkpoint-2")

# Add this before your training loop
for epoch in range(2):
    # This line is crucial for consistent shuffling
    if hasattr(dataloader, 'sampler') and hasattr(dataloader.sampler, 'set_epoch'):
        dataloader.sampler.set_epoch(epoch)
    
    for step, batch in enumerate(dataloader):
        # Your training code here
        print(f"Epoch {epoch}, Step {step}, Batch {batch}")
```

The `set_epoch` call tells the sampler which epoch you're starting, which helps it maintain consistent random shuffling across runs.

Also, a small tip - you might want to use `accelerator.device` instead of hardcoding `'cuda:0'` to make your code more portable:

```python
device = accelerator.device
output = model(torch.tensor([1.0], device=device))
```

Give these changes a try and see if they help with the consistency issue. The data order should now be the same between your original run and the checkpoint-resumed run.

Let me know how it goes or if you run into any other issues!

**Comment by Ildar-Gaisin** [comment]
It seems now that one should replace 

```python
if hasattr(dataloader, 'sampler') and hasattr(dataloader.sampler, 'set_epoch'):
    dataloader.sampler.set_epoch(epoch)
```

with 

```python
if hasattr(dataloader, "set_epoch"):
    dataloader.set_epoch(epoch)
```

perhaps due to an API change? In any case, adding this line has no effect and the batches produced are still in a different order....
